### ISSUE TITLE
reference instances return REFERENCECLASS property in test_client

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbem#598]
in the test case referenceinstance1 (part of pr589) we are getting the following error on the
response:

  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/cim_obj.py", line 498, in _intended_value
    % (name, actual, intended))
ValueError: Property u'Member' has a CIMInstanceName value with classname=u'PyWBEM_Person', but specifies reference_class=u'CIM_ManagedElement' (must be u'PyWBEM_Person')

Not yet sure what this means.  This works in run_cim_operations and I think that CIM_ManagedElement is the correct response so we may actually have an issue in cim_obj.py but not sure yet.

I wanted to gather as much data as possible now.
I think I have isolated the problems. See reasons in following comment block

**Comment by andy-maier** [comment]
This is a check in `CIMProperty`, and we need to investigate as to whether this check is correct and the input data incorrect, vs. the check is overdoing it.

**Comment by KSchopmeyer** [comment]
Agreed. I  moved this one back to mandatory since if it is an issue, it is a real issue.  Right now it is affecting our tests but want to clarify so we can fix tests. 

As an example:

OpenReferences yaml test on the PyWBEM_Person class.

In the test (recorder output) we set the reference_class to CIM_ManagedElement based on the response from OpenPegasus.

The mof for the reference class is:
```
class PyWBEM_MemberOfPersonCollection : CIM_MemberOfCollection {
};
```
and CIM_MemberOfCollection is:

```
class CIM_MemberOfCollection
{
    
        [Key, 
        Aggregate, 
        Description ( "The Collection that aggregates members." )]
    CIM_Collection REF Collection;
    
        [Key, 
        Description ( "The aggregated member of the Collection." )]
    CIM_ManagedElement REF Member;
};
```

The mof generated by the recorder has the property member as follows:

```
      </PROPERTY.REFERENCE>
      <PROPERTY.REFERENCE NAME="Member"  REFERENCECLASS="CIM_ManagedElement">
      <VALUE.REFERENCE>
```
So it would appear to me that the reference class is CIM_ManagedElement, not PyWBEM_Person.

From DSP2001

```
5.2.9 ReferenceClass 625
The ReferenceClass entity is a convenient shorthand for the REFERENCECLASS attribute. If this entity is 626 present, the value shall be a legal CIM class name (DSP0004). 627
Representation of CIM in XML DSP0201
20 DMTF Standard Version 2.4.0
<!ENTITY % ReferenceClass "REFERENCECLASS CDATA #IMPLIED"> 628
The value defines the class name for the reference, and the requirement for the existence of this attribute 629 depends on the element in which it is used. The expected behavior is that the REFERENCECLASS 630 attribute shall exist for classes and should not exist for instances.
```

and

```
The REFERENCECLASS attribute (defined by the ReferenceClass entity), if present, defines the strong 1137 type of the reference. The absence of this attribute indicates that this reference is not strongly typed. The 1138
DSP0201 Representation of CIM in XML
Version 2.4.0 DMTF Standard 33
expected behavior is that the REFERENCECLASS attribute shall exist for PROPERTY.REFERENCE 1139 usage in class entities and should not exist for instance entities because the reference class name should 1140 be defined in the property value.
```

However, what bothers me is that this works when tested against the OpenPegasus server but it caught in a test in cim_obj.py(CIMProperty) in the mock test.

**Comment by KSchopmeyer** [comment]
We moved this to mandatory to at least force us to do the investigation for this release

**Comment by KSchopmeyer** [comment]
The  difference I have been able to find so far is that while the xml appears to be about the same there is one difference: The line  from the yaml
```           
<PROPERTY.REFERENCE NAME="Member" REFERENCECLASS="CIM_ManagedElement">
```
Appears as two lines in the pegasus response
```
              <PROPERTY.REFERENCE NAME="Member" 
                   REFERENCECLASS="CIM_ManagedElement">
```

And the response call to CIMProperty that results does not include a value for the REFERENCE Class

run_pegasus
```
CIMProperty 2264 name=Member value=PyWBEM_Person.CreationClassName="PyWBEM_Person",
	Name="Alice" type=reference class_origin=None reference_class=None
```

yaml

```
CIMProperty 2264 name=Member value=PyWBEM_Person.creationclassname="PyWBEM_Person",
	name="Alice" type=reference class_origin=None reference_class=CIM_ManagedElement
```

it is obviously this last item in the CIMProperty construcor that is causing the problem. Still do not know why and not sure why it is complaining.

Note that OpenPegasus may have a problem here because the spec states that this should not appear for instances.

**Comment by KSchopmeyer** [comment]
I identified several problems:

1. The tuple parser creates a CIMProperty and then adds some of the attributes including the class_reference attribute.  The CIMProperty validator is in __init__ so attributes added after the property is created do not get validated.

```
    pref = CIMProperty(attributes['NAME'], value, type='reference')
    for qual in list_of_matching(tup_tree, ['QUALIFIER']):
        pref.qualifiers[qual.name] = qual
     if 'REFERENCECLASS' in attributes:
        pref.reference_class = attributes['REFERENCECLASS']
    if 'CLASSORIGIN' in attributes:
        pref.class_origin = attributes['CLASSORIGIN']
...
```

2. However, test_client creates the CIMProperty including the reference_class all at once so that it goes through  the __init__(...) validation in which the  intended value function is used to determine if the class_reference property is correct.. That is where it thinks that the intended value should be  PywbemPerson and not CIM_ManagedElement and fails.  Thus the actual failure comes from test_client, not from the received response.

3. The test itself makes no sense for instances in that there is no way for the client to actually know what the reference_class property value should be other than that it is a class name.  It is testing the class defined in the reference_class attribute of the Property against the the class level 

In this case the request is:

```
References for object PyWBEM_Person."alice" role="member",
      ResultClass=WBEM_MemberOfPersonCollection
```
With HTTP request of:
```
<CIM CIMVERSION="2.0" DTDVERSION="2.0">
  <MESSAGE ID="1001" PROTOCOLVERSION="1.0">
    <SIMPLEREQ>
      <IMETHODCALL NAME="References">
        <LOCALNAMESPACEPATH>
          <NAMESPACE NAME="root"/>
          <NAMESPACE NAME="cimv2"/>
        </LOCALNAMESPACEPATH>
        <IPARAMVALUE NAME="ObjectName">
          <INSTANCENAME CLASSNAME="PyWBEM_Person">
            <KEYBINDING NAME="CreationClassName">
              <KEYVALUE VALUETYPE="string">PyWBEM_Person</KEYVALUE>
            </KEYBINDING>
            <KEYBINDING NAME="Name">
              <KEYVALUE VALUETYPE="string">Alice</KEYVALUE>
            </KEYBINDING>
          </INSTANCENAME>
        </IPARAMVALUE>
        <IPARAMVALUE NAME="ResultClass">
          <CLASSNAME NAME="PyWBEM_MemberOfPersonCollection"/>
        </IPARAMVALUE>
        <IPARAMVALUE NAME="Role">
          <VALUE>member</VALUE>
        </IPARAMVALUE>
      </IMETHODCALL>
    </SIMPLEREQ>
  </MESSAGE>
</CIM>
```

the response property 'member' constructor is:

```
CIMProperty 2264
    name=Member
    value=CIMInstanceName(classname=u'PyWBEM_Person',
    keybindings=NocaseDict({u'CreationClassName': u'PyWBEM_Person', u'Name': u'Alice'}),
    namespace=None, host=None) type=reference class_origin=None reference_class=None
```

4. Finally,  the DSP 0201 specification states the following:
       The expected behavior is that the REFERENCECLASS attribute shall exist  for
       PROPERTY.REFERENCE usage in class entities and should not exist for
       instance entities because the reference class name should  be defined in the property value.

so one could argue that pegasus probably should not be passing it back to the  client and the pywbem client should not be using it in the CIMProperty for CIMinstances even if it is received in the xml.

Note that this may be a bit messy because we do not  carry information about whether the creation is for an instance or class between the creation of instances and of their properties. Further, we blindly carry the reference_class information forward from the response into the yaml for with the test_recorder.

We can easily fix the tests_client by changing the xml and pywbem response information in the referenceinstances1 test but that begs the more general problem to fix this completely.

**Comment by KSchopmeyer** [comment]
We agreed that there is a bug.

Andy will fix the cim_obj code and tupleparse.

Karl will fix the tests after the initial submission of the pr with the cim_code, etc.

**Comment by andy-maier** [comment]
My part of the fix is in PR #731. Please add your commits to that PR, because otherwise it will not pass the test_client tests.
